### PR TITLE
feat(policy): reconcile volume policy per volume bases

### PR DESCRIPTION
**Pull request by prateekpandey14**
**What this PR does**:
changes enhances the cvc controller to reconcile the volume target deployment based on the changes done in cstorvolumeconfig (cvc) resources under `Spec.Policy` for individual volumes.

changes reconciled for 
- Tolerations
- Resource Limits
- AuxResoource Limits
- PriorityClass
- QueueDepth
- IOWorkers

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

**Checklist**
- [ ] PR messages has document related information
- [x] Labelled this PR & related issue with `documentation` tag
- [ ] PR messages has breaking changes related information
- [ ] PR messages has upgrade related information
- [ ] Labelled this PR & related issue with `requires-upgrade` tag
- [x] Tests updated

Signed-off-by: prateekpandey14 <prateek.pandey@mayadata.io>

### CODE REVIEW

File: pkg/controllers/cstorvolumeconfig/policy.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+func defaultReplicaPolicy(policy *apis.CStorVolumePolicySpec, defaultPolicy apis.CStorVolumePolicySpec) {
+}
+
+// getDefaultPolicySpec generate default cstor volume policy spec.
+func getDefaultPolicySpec() apis.CStorVolumePolicySpec {
+	return apis.CStorVolumePolicySpec{
+		Target: apis.TargetSpec{
+			Resources: &corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{
+					corev1.ResourceCPU:    resource.MustParse("0"),
```
Comment (mittachaitu) [inline pkg/controllers/cstorvolumeconfig/policy.go:75]: Will this `0` values will be populated into target deployment?

File: pkg/controllers/cstorvolumeconfig/policy.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+					corev1.ResourceMemory: resource.MustParse("0"),
```
Comment (mittachaitu) [inline pkg/controllers/cstorvolumeconfig/policy.go:76]: If this is populated to target deployment then K8s will allocate `0` memory to the pod?

File: pkg/controllers/cstorvolumeconfig/policy.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+// defaultTargetPolicy configure the default volume target deployment related policies
+func defaultTargetPolicy(policy *apis.CStorVolumePolicySpec, defaultPolicy apis.CStorVolumePolicySpec) {
+	if policy.Target.Resources == nil {
+		policy.Target.Resources = defaultPolicy.Target.Resources
+	}
+	if policy.Target.AuxResources == nil {
+		policy.Target.AuxResources = defaultPolicy.Target.AuxResources
+	}
+	if policy.Target.IOWorkers == 0 {
+		policy.Target.IOWorkers = defaultPolicy.Target.IOWorkers
```
Comment (mittachaitu) [inline pkg/controllers/cstorvolumeconfig/policy.go:57]: Do we really need to set it?

File: pkg/controllers/cstorvolumeconfig/policy.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+	// defaultIOWorkers represents default (luWorker) number of threads that are
+	// working on queue
+	defaultIOWorkers = int64(6)
+)
+
+type policyOptFuncs func(*apis.CStorVolumePolicySpec, apis.CStorVolumePolicySpec)
+
+// validatePolicySpec validates the provided policy created by the user and
+// otherwise sets the defaults policy spec for cstor volumes.
+func validatePolicySpec(policy *apis.CStorVolumePolicySpec) {
```
Comment (mittachaitu) [inline pkg/controllers/cstorvolumeconfig/policy.go:38]: `validatePolicySpec` means it should validate right? but it is doing defaulting...

File: pkg/controllers/cstorvolumeconfig/deployment.go [hunk 5]
```diff
@@ -198,35 +199,50 @@ func getTargetMgmtMounts() []corev1.VolumeMount {
 	return defaultMounts
 }
 
+// setIstgtEnvs sets the target container performance tunables env required for
+// cstorvolume target deployment
+func setIstgtEnvs(policySpec *apis.CStorVolumePolicySpec) []corev1.EnvVar {
```
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply pkg/controllers/cstorvolumeconfig/deployment.go:204]: 0 means no resource limits has been set

File: pkg/controllers/cstorvolumeconfig/policy.go (hunk 1 above)
Comment (prateekpandey14) [reply pkg/controllers/cstorvolumeconfig/policy.go:75]: 0 means no resource limits has been set, its like default without any resource limits

File: pkg/controllers/cstorvolumeconfig/policy.go (hunk 2 above)
Comment (prateekpandey14) [reply pkg/controllers/cstorvolumeconfig/policy.go:76]: No , responded in above comment

File: pkg/controllers/cstorvolumeconfig/policy.go (hunk 4 above)
Comment (prateekpandey14) [reply pkg/controllers/cstorvolumeconfig/policy.go:38]: Will be adding more validation based on the invalid value later , hence the name

File: pkg/controllers/cstorvolumeconfig/deployment.go [hunk 6]
```diff
@@ -397,3 +354,77 @@ func (c *CVCController) getOrCreateCStorTargetDeployment(
...
+				WithAnnotationsNew(getDeployTemplateAnnotations()).
+				WithServiceAccountName(util.GetServiceAccountName()).
+				// TODO use of affinity
+				//WithAffinity(getDeployTemplateAffinity()).
+				WithPriorityClassName(getPriorityClass(policySpec)).
+				WithNodeSelectorByValue(policySpec.Target.NodeSelector).
+				WithTolerationsNew(getDeployTolerations(policySpec)...).
+				WithContainers(
+					apicore.NewContainer().
+						WithImage(getVolumeTargetImage()).
```
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply pkg/controllers/cstorvolumeconfig/deployment.go:387]: that's correct , i thought of we are going to stop reconcile via upgrade job set cvc annotation to stop reconciling during the time of upgrade.

File: pkg/controllers/cstorvolumeconfig/controller.go [hunk 7]
```diff
@@ -325,13 +332,62 @@ func (c *CVCController) createVolumeOperation(cvc *apis.CStorVolumeConfig) (*api
 	return cvc, nil
 }
 
+func (c *CVCController) syncPolicySpec(cvc *apis.CStorVolumeConfig) error {
+	cvcCopy := cvc.DeepCopy()
+
+	err := c.patchTargetDeploymentSpec(cvcCopy)
```
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply pkg/controllers/cstorvolumeconfig/controller.go:338]: yes , i am working on that checks right now

File: pkg/controllers/cstorvolumeconfig/deployment.go [hunk 8]
```diff
@@ -397,3 +354,77 @@ func (c *CVCController) getOrCreateCStorTargetDeployment(
...
+			appsv1.RecreateDeploymentStrategyType,
+		).
+		WithSelectorMatchLabelsNew(getDeployMatchLabels(vol.Name)).
+		WithPodTemplateSpec(
+			apicore.NewPodTemplateSpec().
+				WithLabelsNew(getDeployTemplateLabels(vol.Name)).
+				WithAnnotationsNew(getDeployTemplateAnnotations()).
+				WithServiceAccountName(util.GetServiceAccountName()).
+				// TODO use of affinity
+				//WithAffinity(getDeployTemplateAffinity()).
```
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply pkg/controllers/cstorvolumeconfig/deployment.go:381]: Not now,as it not required currently.  we will add them later as we are not supporting the affinity during the creation time itself.

File: pkg/controllers/cstorvolumeconfig/controller.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply pkg/controllers/cstorvolumeconfig/controller.go:338]: trying to do this via creating a hash of the resource and store in labels and compare it once the reconcile request comes. 
Other way is to use `DeepEqual` which i had tested earlier already , but did added into the PR, as its not full proof always as there is numer of mutations happens for the resource once its get created

File: pkg/controllers/cstorvolumeconfig/policy.go [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+				Requests: corev1.ResourceList{
+					corev1.ResourceCPU:    resource.MustParse("0"),
+					corev1.ResourceMemory: resource.MustParse("0"),
+				},
+				Limits: corev1.ResourceList{
+					corev1.ResourceCPU:    resource.MustParse("0"),
+					corev1.ResourceMemory: resource.MustParse("0"),
+				},
+			},
+			QueueDepth: defaultQueueDepth,
```
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply pkg/controllers/cstorvolumeconfig/policy.go:93]: We did set the defaults earlier, so i think its ok to set

File: pkg/controllers/cstorvolumeconfig/policy.go (hunk 3 above)
Comment (prateekpandey14) [reply pkg/controllers/cstorvolumeconfig/policy.go:57]: we did set the defaults earlier , so we can set it

File: pkg/controllers/cstorvolumeconfig/deployment.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply pkg/controllers/cstorvolumeconfig/deployment.go:204]: b/c we have default values then its fine

File: pkg/controllers/cstorvolumeconfig/deployment.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply pkg/controllers/cstorvolumeconfig/deployment.go:387]: Its AntiPattern with respect to controllers :smile: as reconcilation should be idempotent .
But yes we have to do b/c upgrade is not part of controller itself

File: pkg/controllers/cstorvolumeconfig/controller.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/controllers/cstorvolumeconfig/controller.go:338]: I have an idea for this we can discuss and having separate PR also fine

File: pkg/controllers/cstorvolumeconfig/deployment.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply pkg/controllers/cstorvolumeconfig/deployment.go:387]: added the version checks

File: pkg/controllers/cstorvolumeconfig/controller.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply pkg/controllers/cstorvolumeconfig/controller.go]: added the hash changes

File: pkg/util/hash/hash_test.go [hunk 10]
```diff
@@ -124,3 +128,104 @@ func TestHashOutput(t *testing.T) {
...
+
+	// hashes are consistent
+	hash := HashObject(pod)
+	// same object
+	require.Equal(t, hash, HashObject(pod))
+	// different object but same content
+	require.Equal(t, hash, HashObject(samePod))
+
+	// /!\ hashing an object and its pointer lead to different values
+	require.NotEqual(t, hash, HashObject(&pod))
```
Comment (mittachaitu) [inline pkg/util/hash/hash_test.go:191]: IMO it should see only the values it shouldn't bother about pointers/values...

Comment (prateekpandey14) [reply pkg/util/hash/hash_test.go:191]: that is just the test to differentiate the behavior and help us to understand better how the hash is working here , so its good to have. however its a test

**Pull request merged**